File: sessions/2018/easy/637_average_of_levels_in_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        d = {}
        q = [(root, 1)]
        result = []
        while len(q) > 0:
            node, l = q.pop()
            if node:
                if l not in d:
                    d[l] = [node.val]
                else:
                    d[l].append(node.val)
            if node and node.left:
                q.append((node.left, l+1))
            if node and node.right:
                q.append((node.right, l+1))
        for k in sorted(d.keys()):
            result.append(sum(d[k])/float(len(d[k])))
        return result

File: pia/2020/easy/637_average_of_levels_in_binary_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function (root) {
  let list = [root];
  let temp = [];
  let result = [];

  let count = 0;
  let sum = 0;
  while (list.length > 0) {
    const node = list.shift();

    sum += node.val;
    count++;

    if (node.left) {
      temp.push(node.left);
    }
    if (node.right) {
      temp.push(node.right);
    }
    if (list.length === 0) {
      result.push(sum / count);
      sum = 0;
      count = 0;
      list = temp;
      temp = [];
    }
  }
  return result;
};
